[DataLoader] Add simple `groupby` DataPipe

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60840 [WIP] Example of DataPipes and DataFrames integration
* #60680 [DataLoader] Add notebook with DataPipes API example
* #60679 [DataLoader] Add Example Only `fork` DataPipe
* #60678 [DataLoader] Decorate ZipDataPipe as `zip`
* **#60675 [DataLoader] Add simple `groupby` DataPipe**

Differential Revision: [D29461082](https://our.internmc.facebook.com/intern/diff/D29461082)

VitalyFedyunin committed:
[DataLoader] Add simple `groupby` DataPipe

[ghstack-poisoned]

VitalyFedyunin commented:
Entire stack is cleaner attempt to merge/rebase #59053

ejguan commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -205,6 +205,84 @@ def cmp_fn(a : Tuple[str, Any], b : Tuple[str, Any]):
+                buffer_size -= len(buffer[key])
+                del buffer[key]
+
+        while buffer_size:
+            biggest_key = None
+            biggest_size = 0
+            for key in buffer.keys():
+                if len(buffer[key]) > biggest_size:
+                    biggest_size = len(buffer[key])
+                    biggest_key = key
+
+            if self.guaranteed_group_size is not None and biggest_size < self.guaranteed_group_size and not self.drop_remaining:
+                raise Exception('Failed to group items', str(buffer[biggest_key]))
+
+            if self.guaranteed_group_size is None or biggest_size >= self.guaranteed_group_size:
+                yield buffer[biggest_key]
```
It seems we can have a function to return `biggest_key` to reduce the duplicated code.

ejguan commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -205,6 +205,84 @@ def cmp_fn(a : Tuple[str, Any], b : Tuple[str, Any]):
+                        biggest_key = findkey
+
+                if self.guaranteed_group_size is not None and biggest_size < self.guaranteed_group_size and not self.drop_remaining:
+                    raise Exception('Failed to group items', str(buffer[biggest_key]))
```
nit: `RuntimeError` here?

ejguan reviewed (approved):
LGTM with a few nit comments.

VitalyFedyunin committed:
Update on "[DataLoader] Add simple `groupby` DataPipe"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Add simple `groupby` DataPipe"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Add simple `groupby` DataPipe"




[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461082).

VitalyFedyunin committed:
Update on "[DataLoader] Add simple `groupby` DataPipe"


Differential Revision: [D29461082](https://our.internmc.facebook.com/intern/diff/D29461082)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461082).

VitalyFedyunin committed:
Update on "[DataLoader] Add simple `groupby` DataPipe"


Differential Revision: [D29461082](https://our.internmc.facebook.com/intern/diff/D29461082)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461082).

VitalyFedyunin committed:
Update on "[DataLoader] Add simple `groupby` DataPipe"


Differential Revision: [D29461082](https://our.internmc.facebook.com/intern/diff/D29461082)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461082).
